### PR TITLE
Add support for ARM architecture for Singularity build (#286)

### PR DESCRIPTION
**Pull request by pditommaso**
This PR adds the support for arm64 for Singularity builds.

### CODE REVIEW

**Comment by pditommaso** [comment]
Note: this is a second attempt for https://github.com/seqeralabs/wave/pull/286. 


Building Singularity container with this branch, the following error is reported 

```
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_respond: symbol not found
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_alloc: symbol not found
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_free: symbol not found
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_receive: symbol not found
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_fd: symbol not found
Error relocating /usr/local/singularity/bin/singularity: seccomp_notify_id_valid: symbol not found
```

**Comment by marcodelapierre** [comment]
> Building Singularity container with this branch, the following error is reported

This is a build running in an ARM64 machine, right?  What type of instance are we using?

**Comment by pditommaso** [comment]
It uses `m6g.xlarge`. But I think it a side effect of using Singularity 4.0.x

**Comment by munishchouhan** [comment]
> m6g.xlarge

The build was successful in my mac m2

**Comment by pditommaso** [comment]
Likely Singularity 4 has different requirements. Is Singularity 4 really needed for this feature or it could even work with the previous version?

**Comment by munishchouhan** [comment]
> Likely Singularity 4 has different requirements. Is Singularity 4 really needed for this feature or it could even work with the previous version?

we are not using any flag, so singularity 4 is not needed 
we can use previous one also

**Comment by munishchouhan** [comment]
Should i revert to the old version?

**Comment by pditommaso** [comment]
If Singularity 3.x is capable to build the image for arm64, yes

**Comment by munishchouhan** [comment]
ARM build is successful with singularity 3
Screenshot 2023-11-09 at 11 15 23

**Comment by pditommaso** [comment]
and the resulting image is built for arm64 arch?

**Comment by munishchouhan** [comment]
yes, this is what i get using insect
```
docker run -v .:/tmp quay.io/singularity/singularity:v3.11.4-slim-arm64 inspect /tmp/image.sif
org.label-schema.build-arch: arm64
org.label-schema.build-date: Thursday_9_November_2023_10:14:44_UTC
org.label-schema.schema-version: 1.0
org.label-schema.usage.singularity.deffile.bootstrap: docker
org.label-schema.usage.singularity.deffile.from: mambaorg/micromamba:1.5.1
org.label-schema.usage.singularity.version: 3.11.4
org.opencontainers.image.created: 2023-11-06T07:29:22.480Z
org.opencontainers.image.description: Rapid builds of small Conda-based containers using micromamba.
org.opencontainers.image.licenses: Apache-2.0
org.opencontainers.image.revision: 558c7786659753fbbd0ee4c3cd90a7a8b7e2be95
org.opencontainers.image.source: https://github.com/mamba-org/micromamba-docker
org.opencontainers.image.title: micromamba-docker
org.opencontainers.image.url: https://github.com/mamba-org/micromamba-docker
org.opencontainers.image.version: latest
```

**Comment by munishchouhan** [comment]
But there is one issue: i was not able to pull this image using singularity 3.11.4,  i used singularity 4 to pull 

```
munish.chouhan@192 singularityarm % docker run -v .:/tmp quay.io/singularity/singularity:v3.11.4-slim-arm64 pull /tmp/image.sif  oras://hrma017/dev:pandas--eff7f48f310efa1d
time="2023-11-09T10:29:08Z" level=info msg="trying next host" error="failed to do request: Head \"https://hrma017/v2/dev/manifests/pandas--eff7f48f310efa1d\": dial tcp: lookup hrma017 on 192.168.65.7:53: no such host" host=hrma017
FATAL:   While pulling image from oci registry: error fetching image to cache: failed to get checksum for oras://hrma017/dev:pandas--eff7f48f310efa1d: while resolving reference: failed to do request: Head "https://hrma017/v2/dev/manifests/pandas--eff7f48f310efa1d": dial tcp: lookup hrma017 on 192.168.65.7:53: no such host
munish.chouhan@192 singularityarm % docker run -v .:/tmp quay.io/singularity/singularity:v4.0.1-slim-arm64 pull /tmp/image.sif  oras://hrma017/dev:pandas--eff7f48f310efa1d
INFO:    Downloading oras image
```

**Comment by pditommaso** [comment]
There's no option in the `pull` command line to specify the arch/platform?

**Comment by munishchouhan** [comment]
> There's no option in the `pull` command line to specify the arch/platform?
there is --arch flag, but no effect
```
munish.chouhan@192 singularityarm % docker run -v .:/tmp quay.io/singularity/singularity:v4.0.1-slim-arm64 pull --arch arm64 /tmp/image.sif  oras://hrma017/dev:pandas--eff7f48f310efa1d
WARNING: Pull from oras:// is a direct download, --arch and --platform have no effect.
INFO:    Downloading oras image
```

**Comment by pditommaso** [comment]
To summarise to build and push for arm Singularity 3.x is OK. TO pull it it's needed Singularity 4?

**Comment by munishchouhan** [comment]
> To summarise to build and push for arm Singularity 3.x is OK. TO pull it it's needed Singularity 4? is it correct?

Correct

**Comment by munishchouhan** [comment]
> On which machine are you testing the pull ?
I am testing everything on mac m2

>And a question for me, to get context: why are you pulling using an ORAS URL? Does Singularity 3.11 have enough support for it?
What is the other to pull?

File: src/main/groovy/io/seqera/wave/service/builder/KubeBuildStrategy.groovy [hunk 1]
```diff
@@ -117,6 +120,20 @@ class KubeBuildStrategy extends BuildStrategy {
...
 
+    protected String getBuildImage(BuildRequest buildRequest){
+        if( buildRequest.formatDocker() ) {
+            return kanikoImage
+        }
+
+        if( buildRequest.formatSingularity() ) {
+            return buildRequest.platform.arch == "arm64"
+                ? singularityImageArm64
+                :  singularityImage
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/service/builder/KubeBuildStrategy.groovy:131]:
```suggestion
                : singularityImage
```

**Pull request merged**